### ISSUE TITLE
Enable client-side required indicator

### ISSUE THREAD

**Issue by trenzer** [jwaliszko/ExpressiveAnnotations#139]
I might've missed something, but don't know how to indicate required fields, client-side. For instance, show an asterisk on a label for a text input that is required. I got it working only for 'static' required fields, which have the attribute data-val-required.
When the ExpressiveAnnotations attribute 'data-val-requiredif' has been set, I'd like to show or hide the asterisk on the corresponding label, depending on the condition. 

This is what I have for fields that are always required:
 ```
 $('input[type=text], select').each(function () {
                if (undefined != $(this).attr('data-val-required');) {
                    var label = $('label[for="' + $(this).attr('id') + '"]');
                    var text = label.text();
                    var asterisk = '<span class="asterisk">* </span>';
                    if (text.length > 0) {
                        label.prepend(asterisk);
                    }
                }
            });
```

**Comment by jwaliszko** [comment]
You need to trigger the element validation first, in order to know the expression result. Based on the result you know whether the requirement condition is satisfied, so you know if the asterisk should be shown.

**Comment by trenzer** [comment]
Thanks for your quick reply, I now see my question wasn't completely accurate: I want to always show the asterisk when a field is required, based on a condition. Even if a valid value has been entered, I want to keep the indicator.
I looked into triggering eavalid but this seems to only provide the type, whether the condition has been met (valid) and the expression. Is there an easy way to find out whether the field is required? Or do I need to resolve the expression myself to find out whether the field is required?

**Comment by jwaliszko** [comment]
I do not understand why you'd like to evaluate the expression by yourself - isn't the information provided to `eavalid` enough?
```
$('form').find('input, select, textarea').on('eavalid', function(e, type, valid, expr) {
    if (type === 'requiredif') {
        if (valid) { /* field e.currentTarget.name is required */ }
        else { /* field e.currentTarget.name is not required */ }
    }
});
```

**Comment by trenzer** [comment]
Valid doesn't seem to indicate whether e.currentTarget.name is required, valid seems to indicate that the condition was met. :)  On my viewmodel, I have this attribute on property Frequency:  [RequiredIf("Status > 1")]
If I console.log your code with Status = 2 - so that Frequency is required - then parameter valid is false when Frequency is empty. When I enter a value for Frequency, valid is true.  This seems like 'valid' is indicating that the condition is met. I think this doesn't tell me the Frequency is a required input.

**Comment by jwaliszko** [comment]
OK, I was a bit confused, but I understand now. Unfortunately, currently the `eavalid` only gives you the result of the requirement verification, not the result of the condition, based on which the requirement is being verified.

**Comment by jwaliszko** [comment]
Looking at the implementation, you're interested in the result of [this line only](https://github.com/jwaliszko/ExpressiveAnnotations/blob/9abbfc5bec54f3cd4fab780154b42516cdb6c957/src/expressive.annotations.validate.js#L701), instead of the result of the [entire function](https://github.com/jwaliszko/ExpressiveAnnotations/blob/9abbfc5bec54f3cd4fab780154b42516cdb6c957/src/expressive.annotations.validate.js#L694). The thing is, for the requirement verification, the condition is checked only if the field is not empty - if it is not empty, the verification is redundant since field value is given (even from performance perspective computing of the expression is a waste of time here).

Your use case is new though. I cannot think of any simple help here, despite that one you've mentioned - computing the expression by yourself which is a bit tricky.

Maybe we could define some kind of a flag in the settings, e.g. `resolveExpressionsAlways`, and even if value is provided, requirement expression is computed, and even if value is not provided, assertion condition is computed.

**Comment by rickparrish** [comment]
Yes, it does sound very similar to what I was asking in #121.  And I think the same solution would work -- when the 'eadisplay' event is raised the page could handle that and show/hide the askterisk as necessary instead of showing/hiding the whole question as we were doing.

We made additional customizations to EA before putting it into use, so I don't remember for sure whether [those changes](https://github.com/rickparrish/ExpressiveAnnotations/commit/893c724bf929111c945b1d70c88e3c2f804a891a) were sufficient on their own, or if we ran into problems and had to make further adjustments though.

**Comment by jwaliszko** [comment]
With the commit https://github.com/jwaliszko/ExpressiveAnnotations/commit/304721ac4a890e7088310c8387a9f3608bd0851e I've proposed some changes. If you'd like to test, it please fork the repo, uncomment 2 lines of Home.cshtml, mainly [60](https://github.com/jwaliszko/ExpressiveAnnotations/blob/3ed1844fbb83d5bc8fe6653333cd6899dc674894/src/ExpressiveAnnotations.MvcWebSample/Views/Home/Home.cshtml#L60) and [149](https://github.com/jwaliszko/ExpressiveAnnotations/blob/3ed1844fbb83d5bc8fe6653333cd6899dc674894/src/ExpressiveAnnotations.MvcWebSample/Views/Home/Home.cshtml#L149), and give it a try. I'm also opened for any more elegant solutions.

**Comment by trenzer** [comment]
Thanks for the effort and quick thinking! I'm a github newbie so will try that, probably tomorrow.

**Comment by jwaliszko** [comment]
N/P, you're welcome.
It may be funny, but in order to make it clear for future of what this issue is about (mainly for myself ;), truth table is given below:
```
[RequiredIf(cond)]
datatype? Field { get; set; }
```
|field value provided|     cond     |    valid    |
|--------------------|--------------|-------------|
|         0          |      0       |      1      |
|         0          |      1       |      0      |
|         1          |      0       |      1      |
|         1          |      1       |      1      |

**Comment by jwaliszko** [comment]
Change is shipped with version [2.9.2](https://github.com/jwaliszko/ExpressiveAnnotations/releases/tag/v2.9.2) (published yesterday), and respective examples are built into the web sample project ([entry point here](https://github.com/jwaliszko/ExpressiveAnnotations/blob/f2743f62f477ae706c9f1ec3d0157bb6fc3d1413/src/ExpressiveAnnotations.MvcWebSample/Views/Home/Home.cshtml#L150)).
